<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ScDiagnostics BioC2024 Workshop • scDiagnosticsBioc2024Demo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ScDiagnostics BioC2024 Workshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scDiagnosticsBioc2024Demo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BioC2024_scDiagnostics_Workshop.html">ScDiagnostics BioC2024 Workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AnthonyChristidis/scDiagnosticsBioc2024Demo" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ScDiagnostics BioC2024 Workshop</h1>
                        <h4 data-toc-skip class="author">Anthony
Christidis<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/AnthonyChristidis/scDiagnosticsBioc2024Demo/blob/devel/vignettes/BioC2024_scDiagnostics_Workshop.Rmd" class="external-link"><code>vignettes/BioC2024_scDiagnostics_Workshop.Rmd</code></a></small>
      <div class="hidden name"><code>BioC2024_scDiagnostics_Workshop.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="scdiagnostics-bioc2024-workshop">ScDiagnostics BioC2024 Workshop<a class="anchor" aria-label="anchor" href="#scdiagnostics-bioc2024-workshop"></a>
</h2>
<p>Authors: Anthony Christidis<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>, Andrew Ghazi<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>, Ludwig Geistlinger<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>. <br>
Last modified: 22 July, 2024</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p><code>scDiagnostics</code> is a comprehensive toolkit designed for
the analysis and diagnostics of single-cell RNA sequencing (scRNA-seq)
data. This package provides functionalities for comparing principal
components, visualizing canonical correlation analysis (CCA) outputs,
and plotting cell type-specific MDS and PCA projections, among other
tasks.</p>
</div>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<p>Before attending this workshop, participants should have a basic
understanding of single-cell RNA sequencing (scRNA-seq) data analysis
and familiarity with Bioconductor packages. Specifically, knowledge of
the <code>SingleCellExperiment</code> package will be useful for working
with single-cell datasets. While familiarity with dimensionality
reduction techniques, such as Principal Component Analysis (PCA), and
statistical tests will enhance the learning experience, it is not
required.</p>
<p>A solid foundation in single-cell analysis techniques and tools will
significantly benefit those looking to fully leverage the capabilities
of specialized packages like <code>scDiagnostics</code>.” Basic
proficiency in R and R Markdown is also recommended to follow along with
the workshop exercises.</p>
</div>
</div>
<div class="section level3">
<h3 id="setup-r-bioconductor">Setup (<em>R</em> / <em>Bioconductor</em>)<a class="anchor" aria-label="anchor" href="#setup-r-bioconductor"></a>
</h3>
<p>First, let’s load the necessary libraries and data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ccb-hms/scDiagnostics" class="external-link">scDiagnostics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></code></pre></div>
<p>We will use the following datasets available in
<code>scDiagnostics</code> for the workshop. These datasets are
processed versions of the dataset
<code>HeOrganAtlasData(tissue = c("Marrow"))</code> in the Bioconductor
package <code>scRNAseq</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"reference_data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"query_data"</span><span class="op">)</span></span></code></pre></div>
<p>We will also create a dataset where we extract only one cell type
(CD4) from the data. We use the SingleR annotation to subset the data.
THis will allow us to evaluate how well aligned the CD4 cells are
between the reference and query datasets.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset data for CD4</span></span>
<span><span class="va">reference_data_subset</span> <span class="op">&lt;-</span> <span class="va">reference_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">reference_data</span><span class="op">$</span><span class="va">expert_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="va">query_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">query_data</span><span class="op">$</span><span class="va">SingleR_annotation</span> <span class="op">==</span> <span class="st">"CD4"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Selecting highly variable genes</span></span>
<span><span class="va">ref_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">reference_data_subset</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">query_top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">query_data_subset</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Intersect the gene symbols to obtain common genes</span></span>
<span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">ref_top_genes</span>, <span class="va">query_top_genes</span><span class="op">)</span></span>
<span><span class="va">reference_data_subset</span> <span class="op">&lt;-</span> <span class="va">reference_data_subset</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="va">query_data_subset</span><span class="op">[</span><span class="va">common_genes</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Run PCA</span></span>
<span><span class="va">reference_data_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">reference_data_subset</span><span class="op">)</span></span>
<span><span class="va">query_data_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">query_data_subset</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to Annotation Diagnostics</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Brief Overview</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Example Workflow</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Questions and Answers</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<p>The goal of this workshop is to provide participants with a thorough
understanding of the <code>scDiagnostics</code> package, enabling them
to effectively analyze and interpret single-cell RNA sequencing data.
Participants will gain hands-on experience in utilizing the package’s
diverse functionalities to enhance their single-cell data analysis
workflows.</p>
<p>By the end of this workshop, participants will understand the
capabilities and applications of <code>scDiagnostics</code>, master
visualization techniques for both multiple and single cell types,
calculate correlations and distances between cell types, perform and
interpret statistical tests, detect and visualize anomalies, evaluate
data quality and gene overlap, and utilize additional analytical
tools.</p>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<p>By the end of this workshop, participants will be able to:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Visualize Differences Between Datasets:</strong>
<ul>
<li>Use various plotting functions to compare query and reference
datasets across multiple and single cell types.</li>
</ul>
</li>
<li>
<strong>Evaluate Gene Overlap:</strong>
<ul>
<li>Utilize tools for gene overlap measures.</li>
</ul>
</li>
<li>
<strong>Calculate Correlations and Distances:</strong>
<ul>
<li>Implement functions to calculate correlations and distances between
different cell types.</li>
</ul>
</li>
<li>
<strong>Perform Statistical Tests:</strong>
<ul>
<li>Apply statistical methods to compare groups and assess
significance.</li>
</ul>
</li>
<li>
<strong>Detect Anomalies:</strong>
<ul>
<li>Identify and visualize anomalies in single-cell data.</li>
</ul>
</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="brief-overview-of-scdiagnostics">Brief Overview of scDiagnostics<a class="anchor" aria-label="anchor" href="#brief-overview-of-scdiagnostics"></a>
</h3>
<div class="section level4">
<h4 id="visualize-differences-between-datasets">Visualize Differences Between Datasets<a class="anchor" aria-label="anchor" href="#visualize-differences-between-datasets"></a>
</h4>
<div class="section level5">
<h5 id="plotcelltypepca">
<code>plotCellTypePCA</code><a class="anchor" aria-label="anchor" href="#plotcelltypepca"></a>
</h5>
<p>This function plots the principal components for different cell types
in the query and reference datasets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/plotCellTypePCA.html" class="external-link">plotCellTypePCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                           reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                           cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                           query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                           ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                           pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">pc_plot</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/pca_plot-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="boxplotpca">
<code>boxplotPCA</code><a class="anchor" aria-label="anchor" href="#boxplotpca"></a>
</h5>
<p>This function generates a boxplot visualization of principal
components (PCs) for different cell types across two datasets (query and
reference).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc_boxplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/boxplotPCA.html" class="external-link">boxplotPCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                         reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                         query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                         ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                         cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                         pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pc_boxplot</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/boxplot-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="plotcelltypemds">
<code>plotCellTypeMDS</code><a class="anchor" aria-label="anchor" href="#plotcelltypemds"></a>
</h5>
<p>This function facilitates the assessment of similarity between
reference and query datasets through Multidimensional Scaling (MDS)
scatter plots. It allows the visualization of cell types based on a
dissimilarity matrix computed from a user-selected gene set.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mds_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/plotCellTypeMDS.html" class="external-link">plotCellTypeMDS</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                            reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                            cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span>,</span>
<span>                            query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                            ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span><span class="op">)</span></span>
<span><span class="va">mds_plot</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/mds-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="comparepca">
<code>comparePCA</code><a class="anchor" aria-label="anchor" href="#comparepca"></a>
</h5>
<p>This function compares the principal components (PCs) obtained from
separate PCA on reference and query datasets for a single cell type
using either cosine similarity or correlation.</p>
<p>The S3 plot method generates a heatmap to visualize the cosine
similarities between principal components from the output of the
comparePCA function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">similarity_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/comparePCA.html" class="external-link">comparePCA</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data_subset</span>, </span>
<span>                             reference_data <span class="op">=</span> <span class="va">reference_data_subset</span>, </span>
<span>                             query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                             ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                             pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">similarity_mat</span><span class="op">)</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/pca_comparison-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level4">
<h4 id="evaluate-gene-overlap">Evaluate Gene Overlap<a class="anchor" aria-label="anchor" href="#evaluate-gene-overlap"></a>
</h4>
<div class="section level5">
<h5 id="calculatevarimpoverlap">
<code>calculateVarImpOverlap</code><a class="anchor" aria-label="anchor" href="#calculatevarimpoverlap"></a>
</h5>
<p>This function uses the Random Forest algorithm to calculate the
importance of genes in differentiating between cell types within both a
reference dataset and a query dataset. The function then compares the
top genes identified in both datasets to determine the overlap in their
importance scores.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var_imp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/calculateVarImpOverlap.html" class="external-link">calculateVarImpOverlap</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                    query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                    ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                    query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span><span class="op">)</span></span>
<span><span class="va">var_imp</span><span class="op">[[</span><span class="st">"var_imp_comparison"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;     B_and_plasma-CD8     B_and_plasma-CD4 B_and_plasma-Myeloid </span></span>
<span><span class="co">#&gt;                 0.80                 0.88                 0.34 </span></span>
<span><span class="co">#&gt;              CD8-CD4          CD8-Myeloid          CD4-Myeloid </span></span>
<span><span class="co">#&gt;                 0.70                 0.24                 0.40</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="calculate-correlations-and-distances">Calculate Correlations and Distances<a class="anchor" aria-label="anchor" href="#calculate-correlations-and-distances"></a>
</h4>
<div class="section level5">
<h5 id="calculateaveragepairwisecorrelation">
<code>calculateAveragePairwiseCorrelation</code><a class="anchor" aria-label="anchor" href="#calculateaveragepairwisecorrelation"></a>
</h5>
<p>Computes the average pairwise correlations between specified cell
types in single-cell gene expression data.</p>
<p>The S3 plot method takes the output of the
<code>calculateAveragePairwiseCorrelation</code> function, which should
be a matrix of pairwise correlations, and plots it as a heatmap.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_matrix_avg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/calculateAveragePairwiseCorrelation.html" class="external-link">calculateAveragePairwiseCorrelation</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                                      reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                                      query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                                      ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                                      cell_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"B_and_plasma"</span><span class="op">)</span>, </span>
<span>                                                      pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                                                      correlation_method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cor_matrix_avg</span><span class="op">)</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/pairwise_correlations-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level4">
<h4 id="perform-statistical-tests">Perform Statistical Tests<a class="anchor" aria-label="anchor" href="#perform-statistical-tests"></a>
</h4>
<div class="section level5">
<h5 id="calculatenearestneighborprobabilities">
<code>calculateNearestNeighborProbabilities</code><a class="anchor" aria-label="anchor" href="#calculatenearestneighborprobabilities"></a>
</h5>
<p>This function computes the probabilities for each query cell of
belonging to either the reference or query dataset for each cell type
using nearest neighbor analysis.</p>
<p>The S3 plot method generates a density plot showing the distribution
of probabilities for each cell of belonging to either the reference or
query dataset for each cell type.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nn_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/calculateNearestNeighborProbabilities.html" class="external-link">calculateNearestNeighborProbabilities</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                                   reference_data <span class="op">=</span> <span class="va">reference_data</span>,</span>
<span>                                                   query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                                   ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                                                   pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">nn_output</span><span class="op">)</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/probabilities-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 id="calculatehotellingpvalue">
<code>calculateHotellingPValue</code><a class="anchor" aria-label="anchor" href="#calculatehotellingpvalue"></a>
</h5>
<p>This function calculates Hotelling’s T-squared statistic for
comparing multivariate means between reference and query datasets,
projected onto a subset of principal components (PCs). It performs a
permutation test to obtain p-values for each cell type specified.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/calculateHotellingPValue.html" class="external-link">calculateHotellingPValue</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                     reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                     query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                     ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                                     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">p_values</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; B_and_plasma          CD8          CD4      Myeloid </span></span>
<span><span class="co">#&gt;        0.006        0.000        0.000        0.000</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="detect-anomalies">Detect Anomalies<a class="anchor" aria-label="anchor" href="#detect-anomalies"></a>
</h4>
<div class="section level5">
<h5 id="detectanomaly">
<code>detectAnomaly</code><a class="anchor" aria-label="anchor" href="#detectanomaly"></a>
</h5>
<p>This function projects the query data onto the PCA space of the
reference data. An isolation forest is then built on the reference data
to identify anomalies in the query data based on their PCA projections.
If no query dataset is provided by the user, the anomaly scores are
computed on the reference data itself. Anomaly scores for the data with
all combined cell types are also provided as part of the output.</p>
<p>The S3 plot method extracts the specified PCs from the given anomaly
detection object and generates scatter plots for each pair of PCs. It
uses ggplot2 to create a faceted plot where each facet represents a pair
of PCs. Anomalies are highlighted in red, while normal points are shown
in black.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anomaly_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/detectAnomaly.html" class="external-link">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>, </span>
<span>                                query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>,</span>
<span>                                pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>                                n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                anomaly_treshold <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">anomaly_output</span>, </span>
<span>     cell_type <span class="op">=</span> <span class="st">"CD4"</span>, </span>
<span>     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, </span>
<span>     data_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"query"</span>, <span class="st">"reference"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/anomaly-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="example-workflow">Example Workflow<a class="anchor" aria-label="anchor" href="#example-workflow"></a>
</h3>
<p>First let us see how well the <code>SingleR</code> annotation
performed</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>Expert_Annotation <span class="op">=</span> <span class="va">query_data</span><span class="op">$</span><span class="va">expert_annotation</span>, </span>
<span>      SingleR <span class="op">=</span> <span class="va">query_data</span><span class="op">$</span><span class="va">SingleR_annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  SingleR</span></span>
<span><span class="co">#&gt; Expert_Annotation B_and_plasma CD4 CD8 Myeloid</span></span>
<span><span class="co">#&gt;      B_and_plasma          136   0   0      10</span></span>
<span><span class="co">#&gt;      CD4                     0 190   1      13</span></span>
<span><span class="co">#&gt;      CD8                     0 134 196      29</span></span>
<span><span class="co">#&gt;      Myeloid                 0   0   0      40</span></span></code></pre></div>
<p>We can see that according to expert annotation, a lot of CD8 cells
were incorrectly labeled as CD4 cells by <code>SingleR</code>. Let’s
identify some of these cells (the top 5 deviating cells for CD4).</p>
<p>First, we compute the distances from each query cell to each
reference cell.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">distance_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/calculateCellDistances.html" class="external-link">calculateCellDistances</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                        reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                        query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                        ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                                        pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> </span></code></pre></div>
<p>Now, let’s compute the top six anomalies for CD4 using the anomaly
detection function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cd4_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/detectAnomaly.html" class="external-link">detectAnomaly</a></span><span class="op">(</span>reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                               query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                               query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                               ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                               pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                               n_tree <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                               anomaly_treshold <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">$</span><span class="va">CD4</span></span>
<span><span class="va">cd4_top6_anomalies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">cd4_anomalies</span><span class="op">$</span><span class="va">query_anomaly_scores</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Next, we plot the densities of the distances for each of these
anomalous cells for CD4 and CD8.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD4"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/plot_anomaly_cd4-1.png" width="100%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">distance_data</span>, ref_cell_type <span class="op">=</span> <span class="st">"CD8"</span>, cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span><span class="op">)</span></span></code></pre></div>
<p><img src="BioC2024_scDiagnostics_Workshop_files/figure-html/plot_anomaly_cd4-2.png" width="100%" style="display: block; margin: auto;"></p>
<p>We can also compute some measures of similarity between these cells
and all cell types.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overlap_measures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scDiagnostics/man/calculateCellDistancesSimilarity.html" class="external-link">calculateCellDistancesSimilarity</a></span><span class="op">(</span>query_data <span class="op">=</span> <span class="va">query_data</span>, </span>
<span>                                                     reference_data <span class="op">=</span> <span class="va">reference_data</span>, </span>
<span>                                                     cell_names <span class="op">=</span> <span class="va">cd4_top6_anomalies</span>,</span>
<span>                                                     query_cell_type_col <span class="op">=</span> <span class="st">"SingleR_annotation"</span>, </span>
<span>                                                     ref_cell_type_col <span class="op">=</span> <span class="st">"expert_annotation"</span>,</span>
<span>                                                     pc_subset <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> </span>
<span><span class="va">overlap_measures</span></span>
<span><span class="co">#&gt; $bhattacharyya_coef</span></span>
<span><span class="co">#&gt;                 Cell B_and_plasma       CD8       CD4    Myeloid</span></span>
<span><span class="co">#&gt; 1 AAACGGGCACATGACT-1    0.1925182 0.7128711 0.2116577 0.07436962</span></span>
<span><span class="co">#&gt; 2 ATTGGTGGTAACGCGA-1    0.4236378 0.3799852 0.3830778 0.27169272</span></span>
<span><span class="co">#&gt; 3 ATTGGACCATACTACG-1    0.2097601 0.8765028 0.3656771 0.09452698</span></span>
<span><span class="co">#&gt; 4 CATTCGCTCGTCACGG-1    0.1363147 0.7006176 0.2448559 0.05778528</span></span>
<span><span class="co">#&gt; 5 CACCTTGGTTATGTGC-1    0.2452612 0.9885897 0.4514740 0.04147370</span></span>
<span><span class="co">#&gt; 6 TTGCGTCTCCACTCCA-1    0.2741303 0.9457587 0.5678963 0.12581074</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $hellinger_dist</span></span>
<span><span class="co">#&gt;                 Cell B_and_plasma       CD8       CD4   Myeloid</span></span>
<span><span class="co">#&gt; 1 AAACGGGCACATGACT-1    0.8985999 0.5358441 0.8878864 0.9620969</span></span>
<span><span class="co">#&gt; 2 ATTGGTGGTAACGCGA-1    0.7591853 0.7874102 0.7854439 0.8534092</span></span>
<span><span class="co">#&gt; 3 ATTGGACCATACTACG-1    0.8889544 0.3514217 0.7964439 0.9515635</span></span>
<span><span class="co">#&gt; 4 CATTCGCTCGTCACGG-1    0.9293467 0.5471585 0.8689903 0.9706775</span></span>
<span><span class="co">#&gt; 5 CACCTTGGTTATGTGC-1    0.8687570 0.1068190 0.7406254 0.9790436</span></span>
<span><span class="co">#&gt; 6 TTGCGTCTCCACTCCA-1    0.8519799 0.2328977 0.6573459 0.9349809</span></span></code></pre></div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:anthony-alexander_christidis@hms.harvard.edu" class="email">anthony-alexander_christidis@hms.harvard.edu</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Core for Computational Biomedicine, Harvard Medical
School<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Core for Computational Biomedicine, Harvard Medical
School<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>Core for Computational Biomedicine, Harvard Medical
School<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Anthony Christidis, Andrew Ghazi, Smriti Chawla, Ludwig Geistlinger, Robert Gentleman.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
